### PR TITLE
HA member enters bad state unable to copy store from master

### PR DESCRIPTION
**Pull request by spacecowboy**
If an error occurs while copying files from the temporary directory
to the actual store directory, the slave can end up in a weird unrecoverable
state where the services aren't ever started.

By cleaning up the store directory if an error occurs, copyFromMaster will
be called again on subsequent attempts (it would not be called again
previously).

Then the copying can be retried, and if successful, the services will be started
again.

The PR also moves some code previously located in BranchedDataStore, to avoid
code duplication.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    public static File newBranchedDataDir( File storeDir )
+    {
+        File result = getBranchedDataDirectory( storeDir, System.currentTimeMillis() );
+        result.mkdirs();
+        return result;
+    }
+
+    private static boolean isBranchedDataRootDirectory( File directory )
+    {
+        return directory.isDirectory() && directory.getName().equals( BRANCH_SUBDIRECTORY );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java]: Strange naming, `directory.isDirectory()` looks weird, shouldn't param name be just `file` instead?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public static boolean isBranchedDataDirectory( File directory )
+    {
+        return directory.isDirectory() && directory.getParentFile().getName().equals( BRANCH_SUBDIRECTORY ) &&
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java]: Strange naming, `directory.isDirectory()` looks weird, shouldn't param name be just `file` instead?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        return new File( storeDir, BRANCH_SUBDIRECTORY );
+    }
+
+    public static File getBranchedDataDirectory( File storeDir, long timestamp )
+    {
+        return new File( getBranchedDataRootDirectory( storeDir ), "" + timestamp );
+    }
+
+    private static boolean isAllDigits( String string )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java]: `StringUtils.isNumeric`?

**Comment by spacecowboy** [comment]
@MishaDemianenko Agreed, now changed.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        return storeDir.listFiles( new FileFilter()
+        {
+            @Override
+            public boolean accept( File file )
+            {
+                return !file.getName().startsWith( "metrics" ) && !file.getName().startsWith( "messages." ) &&
+                       !isBranchedDataRootDirectory( file );
+            }
+        } );
+    }
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/StoreUtil.java:57]: Wondering if this is related to `NeoStoreFileListing`

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/BranchedDataPolicy.java [hunk 5]
```diff
@@ -56,92 +61,9 @@ public void handle( File storeDir ) throws IOException
                 @Override
                 public void handle( File storeDir ) throws IOException
                 {
-                    for ( File file : relevantDbFiles( storeDir ) )
-                    {
-                        FileUtils.deleteRecursively( file );
-                    }
+                    cleanStoreDir( storeDir );
```
Comment (davidegrohmann) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/BranchedDataPolicy.java:64]: Can't you use `cleanStoreDir(...)` in `keep_last` as well?

Comment (spacecowboy) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/BranchedDataPolicy.java:64]: `keep_last` deletes the entire directory. `cleanStoreDir`deletes only the "relevant" files inside that directory.

**Pull request merged**